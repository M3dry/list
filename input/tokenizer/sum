(+ 4 1)

-- args: -i -f
-- expected stdout:
-- Ok(
--     Tokens(
--         [
--             ParenOpen,
--             Identifier(
--                 "+",
--             ),
--             Literal(
--                 Int(
--                     false,
--                     4,
--                 ),
--             ),
--             Literal(
--                 Int(
--                     false,
--                     1,
--                 ),
--             ),
--             ParenClose,
--         ],
--     ),
-- )

