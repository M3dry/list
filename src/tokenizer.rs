use std::{
    iter::Peekable,
    str::{Chars, FromStr},
};

#[derive(Debug, PartialEq, Eq)]
pub struct Tokens(pub(crate) Vec<Token>);

#[derive(Debug, PartialEq, Eq)]
pub(crate) enum Token {
    Literal(Literals),
    Keyword(Keywords),
    Type(BuiltinTypes),
    ParenOpen,
    ParenClose,
    BracketOpen,
    BracketClose,
    AngleBracketOpen,
    AngleBracketClose,
    CurlyOpen,
    CurlyClose,
    Identifier(String),
    Generic(String),
}

impl From<Vec<Token>> for Tokens {
    fn from(value: Vec<Token>) -> Self {
        Tokens(value)
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum Literals {
    Int(bool, u128),
    String(String),
    Char(char),
    Bool(bool),
}

impl ToString for Literals {
    fn to_string(&self) -> String {
        match self {
            Literals::Int(true, num) => format!("-{}", &num.to_string()),
            Literals::Int(false, num) => num.to_string(),
            Literals::String(str) => format!(r#""{str}""#),
            Literals::Char(char) => format!("'{char}'"),
            Literals::Bool(true) => "true".to_string(),
            Literals::Bool(false) => "false".to_string(),
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub(crate) enum Keywords {
    If,
    Match,
    Defun,
    Lambda,
    Let,
    Struct,
    Enum,
    Arrow,
}

#[derive(Debug, PartialEq, Eq)]
pub enum BuiltinTypes {
    U8,
    U16,
    U32,
    U64,
    U128,
    I8,
    I16,
    I32,
    I64,
    I128,
    String,
    Char,
    Bool,
}

impl ToString for BuiltinTypes {
    fn to_string(&self) -> String {
        match self {
            BuiltinTypes::U8 => "u8",
            BuiltinTypes::U16 => "u16",
            BuiltinTypes::U32 => "u32",
            BuiltinTypes::U64 => "u64",
            BuiltinTypes::U128 => "k128",
            BuiltinTypes::I8 => "i8",
            BuiltinTypes::I16 => "i16",
            BuiltinTypes::I32 => "i32",
            BuiltinTypes::I64 => "i64",
            BuiltinTypes::I128 => "i128",
            BuiltinTypes::String => "String",
            BuiltinTypes::Char => "char",
            BuiltinTypes::Bool => "bool",
        }
        .to_string()
    }
}

impl FromStr for Tokens {
    type Err = &'static str;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut tokens = vec![];
        let mut chars = s.chars().peekable();

        'l: while let Some(char) = chars.next() {
            match char {
                '(' => tokens.push(Token::ParenOpen),
                ')' => tokens.push(Token::ParenClose),
                '[' => tokens.push(Token::BracketOpen),
                '<' => tokens.push(Token::AngleBracketOpen),
                '>' => tokens.push(Token::AngleBracketClose),
                ']' => tokens.push(Token::BracketClose),
                '{' => tokens.push(Token::CurlyOpen),
                '}' => tokens.push(Token::CurlyClose),
                '\'' => {
                    let ch = chars.next().unwrap();

                    if !matches!(chars.next(), Some('\'')) {
                        return Err("unfinished closing single quote");
                    }

                    tokens.push(Token::Literal(Literals::Char(ch)));
                }
                '"' => {
                    let mut literal = vec![chars.next().unwrap()];
                    let mut backslash = false;

                    while let Some(char) = chars.next() {
                        if char == '"' && !backslash {
                            break;
                        } else if char == '\\' && !backslash {
                            backslash = true
                        } else {
                            literal.push(char);
                            backslash = false
                        }
                    }

                    tokens.push(Token::Literal(Literals::String(
                        literal.into_iter().collect(),
                    )));
                }
                char => {
                    if char == '('
                        || char == ')'
                        || char == '['
                        || char == ']'
                        || char == '<'
                        || char == '>'
                        || char == ' '
                    {
                        continue;
                    }
                    let mut chs = vec![];

                    if char == '-' {
                        match check_arrow(&mut chars)? {
                            Triple::Token(token) => {
                                tokens.push(token);
                                continue 'l;
                            }
                            Triple::Char(char) => {
                                chs.push(char);
                            }
                            Triple::None => {}
                        }
                    } else {
                        chs.push(char)
                    }

                    while let Some(char) = chars.peek() {
                        if *char == '('
                            || *char == ')'
                            || *char == '['
                            || *char == ']'
                            || *char == '<'
                            || *char == '>'
                            || *char == ' '
                        {
                            break;
                        } else if *char == '-' {
                            let ch = chars.next().unwrap();
                            match check_arrow(&mut chars)? {
                                Triple::Token(token) => {
                                    tokens.append(&mut token_from_str(
                                        &chs.into_iter().collect::<String>()[..],
                                    ));
                                    tokens.push(token);
                                    continue 'l;
                                }
                                Triple::Char(char) => {
                                    chs.push(ch);
                                    chs.push(char);
                                }
                                Triple::None => {}
                            }
                        } else {
                            chs.push(chars.next().unwrap())
                        }
                    }

                    tokens.append(&mut token_from_str(
                        &chs.into_iter().collect::<String>()[..],
                    ));
                }
            }
        }

        Ok(Tokens::from(tokens))
    }
}

fn token_from_str(str: &str) -> Vec<Token> {
    match str {
        "if" => vec![Token::Keyword(Keywords::If)],
        "match" => vec![Token::Keyword(Keywords::Match)],
        "defun" => vec![Token::Keyword(Keywords::Defun)],
        "lambda" => vec![Token::Keyword(Keywords::Lambda)],
        "let" => vec![Token::Keyword(Keywords::Let)],
        "struct" => vec![Token::Keyword(Keywords::Struct)],
        "enum" => vec![Token::Keyword(Keywords::Enum)],
        "true" => vec![Token::Literal(Literals::Bool(true))],
        "false" => vec![Token::Literal(Literals::Bool(false))],
        "u8" => vec![Token::Type(BuiltinTypes::U8)],
        "u16" => vec![Token::Type(BuiltinTypes::U16)],
        "u32" => vec![Token::Type(BuiltinTypes::U32)],
        "u64" => vec![Token::Type(BuiltinTypes::U64)],
        "u128" => vec![Token::Type(BuiltinTypes::U128)],
        "i8" => vec![Token::Type(BuiltinTypes::I8)],
        "i16" => vec![Token::Type(BuiltinTypes::I16)],
        "i32" => vec![Token::Type(BuiltinTypes::I32)],
        "i64" => vec![Token::Type(BuiltinTypes::I64)],
        "i128" => vec![Token::Type(BuiltinTypes::I128)],
        "string" => vec![Token::Type(BuiltinTypes::String)],
        "char" => vec![Token::Type(BuiltinTypes::Char)],
        "bool" => vec![Token::Type(BuiltinTypes::Bool)],
        arrow if arrow.len() >= 2 && &arrow[..2] == "->" => {
            let mut ret = vec![Token::Keyword(Keywords::Arrow)];
            ret.append(&mut token_from_str(&str[2..]));
            ret
        }
        num if num.len() > 1 && &num[..1] == "-" && num[1..].parse::<u128>().is_ok() => {
            vec![Token::Literal(Literals::Int(
                true,
                num[1..].parse().unwrap(),
            ))]
        }
        num if num.parse::<u128>().is_ok() => {
            vec![Token::Literal(Literals::Int(false, num.parse().unwrap()))]
        }
        generic if generic.len() > 1 && &generic[..1] == ":" => {
            vec![Token::Generic(generic[1..].to_string())]
        }
        identifier
            if identifier.len() > 0
                && identifier.chars().all(|c| {
                    c.is_ascii_alphanumeric()
                        || [
                            '!', '@', '#', '$', '%', '^', '&', '*', '-', '_', '=', '+', ',', '.',
                            '<', '>', '?',
                        ]
                        .contains(&c)
                }) =>
        {
            vec![Token::Identifier(identifier.to_string())]
        }
        _ => vec![],
    }
}

enum Triple {
    Token(Token),
    Char(char),
    None,
}

fn check_arrow(chars: &mut Peekable<Chars>) -> Result<Triple, &'static str> {
    match chars.next() {
        Some('-') => {
            while !matches!(chars.next(), Some('\n') | None) {}
            Ok(Triple::None)
        }
        Some('>') => Ok(Triple::Token(Token::Keyword(Keywords::Arrow))),
        Some(char) => Ok(Triple::Char(char)),
        _ => Err("Expected more chars"),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/tokenizer/");
                settings.bind(|| {
                    insta::assert_snapshot!(contents
                        .lines()
                        .filter_map(|line| if line != "" {
                            Some(format!("{line}\n{:#?}", Tokens::from_str(line)))
                        } else {
                            None
                        })
                        .collect::<Vec<String>>()
                        .join("\n\n"));
                });
            }
        };
    }

    snapshot!(test_calling, "../testdata/input/calling.lt");
    snapshot!(test_if, "../testdata/input/if.lt");
    snapshot!(test_match, "../testdata/input/match.lt");
    snapshot!(test_defun, "../testdata/input/defun.lt");
    snapshot!(test_lambda, "../testdata/input/lambda.lt");
    snapshot!(test_let, "../testdata/input/let.lt");
    snapshot!(test_struct, "../testdata/input/struct.lt");
    snapshot!(test_enum, "../testdata/input/enum.lt");
}
