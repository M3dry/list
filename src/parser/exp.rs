use crate::tokenizer::{Keywords, Literals, Token};

use super::{
    error, lambda::Lambda, r#if::If, r#let::Let, r#match::Match, r#type::NamespacedType, Parser,
    ParserError, ParserErrorStack,
};

#[derive(Debug)]
pub enum Exp {
    Lambda(Box<Lambda>),
    If(Box<If>),
    Match(Box<Match>),
    Let(Box<Let>),
    FuncCall(Box<Exp>, Vec<Exp>),
    Ref(Box<Exp>),
    MutRef(Box<Exp>),
    Deref(Box<Exp>),
    Negation(Box<Exp>),
    Infix(Box<Infix>),
    // TODO: Do(Box<Do>),
    Variable(String),
    Literal(Literals),
    TypeCreation(TypeCreation),
}

impl TryFrom<&mut Parser> for Exp {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let peek = value
            .first()
            .ok_or(error!("Exp", format!("Expected more tokens")))?;

        match peek {
            Token::Identifier(_) => {
                if matches!(
                    value.nth(1),
                    Some(&Token::Keyword(Keywords::Arrow) | &Token::CurlyOpen)
                ) {
                    Ok(Exp::TypeCreation(error!(
                        TypeCreation::try_from(&mut *value),
                        "Exp"
                    )?))
                } else if let Token::Identifier(iden) = value.pop_front().unwrap() {
                    Ok(Exp::Variable(iden))
                } else {
                    unreachable!()
                }
            }
            Token::Literal(_) => {
                if let Token::Literal(literal) = value.pop_front().unwrap() {
                    Ok(Exp::Literal(literal))
                } else {
                    unreachable!()
                }
            }
            Token::Ref => {
                value.pop_front();

                if value.first() == Some(&Token::Keyword(Keywords::Mut)) {
                    value.pop_front();
                    Ok(Exp::MutRef(Box::new(error!(
                        Exp::try_from(&mut *value),
                        "Exp"
                    )?)))
                } else {
                    Ok(Exp::Ref(Box::new(error!(
                        Exp::try_from(&mut *value),
                        "Exp"
                    )?)))
                }
            }
            Token::Char('*') => {
                value.pop_front();

                Ok(Exp::Deref(Box::new(error!(
                    Exp::try_from(&mut *value),
                    "Exp"
                )?)))
            }
            Token::AngleBracketOpen => Ok(Exp::TypeCreation(error!(
                TypeCreation::try_from(&mut *value),
                "Exp"
            )?)),
            Token::ParenOpen => {
                match value
                    .nth(1)
                    .ok_or(error!("Exp", format!("Expected more tokens")))?
                {
                    Token::Keyword(Keywords::Not) => {
                        value.pop_front();
                        value.pop_front();

                        Ok(Exp::Negation(Box::new(error!(
                            Exp::try_from(&mut *value),
                            "Exp"
                        )?)))
                    }
                    Token::Keyword(Keywords::If) => {
                        Ok(Exp::If(Box::new(error!(If::try_from(&mut *value), "Exp")?)))
                    }
                    Token::Keyword(Keywords::Match) => Ok(Exp::Match(Box::new(error!(
                        Match::try_from(&mut *value),
                        "Exp"
                    )?))),
                    Token::Keyword(Keywords::Let) => Ok(Exp::Let(Box::new(error!(
                        Let::try_from(&mut *value),
                        "Exp"
                    )?))),
                    Token::Keyword(Keywords::Lambda) => Ok(Exp::Lambda(Box::new(error!(
                        Lambda::try_from(&mut *value),
                        "Exp"
                    )?))),
                    Token::Identifier(_) => {
                        if matches!(
                            value.nth(2),
                            Some(&Token::Keyword(Keywords::Arrow) | &Token::CurlyClose)
                        ) {
                            Ok(Exp::TypeCreation(error!(
                                TypeCreation::try_from(&mut *value),
                                "Exp"
                            )?))
                        } else {
                            value.pop_front();
                            let func =
                                if let Token::Identifier(iden) = value.pop_front_err("Exp")? {
                                    iden
                                } else {
                                    unreachable!()
                                };
                            let mut args = vec![];

                            loop {
                                let peek = value
                                    .first()
                                    .ok_or(error!("Exp", format!("Expected more tokens")))?;

                                if peek == &Token::ParenClose {
                                    value.pop_front();
                                    break;
                                }

                                args.push(error!(Exp::try_from(&mut *value), "exp")?)
                            }

                            Ok(Exp::FuncCall(Box::new(Exp::Variable(func)), args))
                        }
                    }
                    _ => {
                        value.pop_front();

                        if let Ok(infix) = Infix::try_from(&mut *value) {
                            let infix = Ok(Self::Infix(Box::new(infix)));
                            let next = value.pop_front_err("Exp")?;
                            if next != Token::ParenClose {
                                return Err(error!(
                                    "Exp",
                                    format!("Expected parenClose, got {next:#?}")
                                ));
                            }
                            return infix;
                        }

                        let exp = error!(Exp::try_from(&mut *value), "Exp")?;

                        let next = value
                            .first()
                            .ok_or(error!("Exp", format!("Expected more tokens")))?;
                        if next == &Token::ParenClose {
                            value.pop_front();
                            Ok(exp)
                        } else {
                            let mut params = vec![];

                            loop {
                                let peek = value
                                    .first()
                                    .ok_or(error!("Exp", format!("Expected more tokens")))?;

                                if peek == &Token::ParenClose {
                                    value.pop_front();
                                    break;
                                }

                                params.push(error!(Exp::try_from(&mut *value), "Exp")?)
                            }

                            Ok(Exp::FuncCall(Box::new(exp), params))
                        }
                    }
                }
            }
            token => {
                panic!("Not implemented {token:#?}")
            }
        }
    }
}

impl ToString for Exp {
    fn to_string(&self) -> String {
        match self {
            Exp::Lambda(lambda) => format!("({})", lambda.to_string()),
            Exp::If(r#if) => format!("{{{}}}", r#if.to_string()),
            Exp::Match(r#match) => format!("{{{}}}", r#match.to_string()),
            Exp::Let(r#let) => format!("{{{}}}", r#let.to_string()),
            Exp::FuncCall(func, args) => format!(
                "{{{}({})}}",
                func.to_string(),
                &if args.is_empty() {
                    format!(", ")
                } else {
                    args.iter().fold(String::new(), |str, arg| {
                        format!("{str}, {}", arg.to_string())
                    })
                }[2..]
            ),
            Exp::Ref(exp) => format!("{{&{{{}}}}}", exp.to_string()),
            Exp::MutRef(exp) => format!("{{&mut {{{}}}}}", exp.to_string()),
            Exp::Deref(exp) => format!("{{*{{{}}}}}", exp.to_string()),
            Exp::Negation(exp) => format!("{{!{{{}}}}}", exp.to_string()),
            Exp::Infix(infix) => format!("{{{}}}", infix.to_string()),
            Exp::Variable(var) => format!("{var}"),
            Exp::Literal(literal) => format!("{}", literal.to_string()),
            Exp::TypeCreation(creation) => format!("{{{}}}", creation.to_string()),
        }
    }
}

#[derive(Debug)]
pub enum TypeCreation {
    Simple(NamespacedType),
    Vars(NamespacedType, Vec<Exp>),
    Struct(NamespacedType, Vec<(String, Exp)>),
    Touple(Vec<Exp>),
}

impl TryFrom<&mut Parser> for TypeCreation {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        let mut paren = false;
        let next = value
            .first()
            .ok_or(error!("TypeCreation", format!("Expected more tokens")))?;
        if next == &Token::ParenOpen {
            value.pop_front();
            paren = true;
        } else if next == &Token::AngleBracketOpen {
            value.pop_front();
            let mut exps = vec![];

            loop {
                let peek = value
                    .first()
                    .ok_or(error!("TypeCreation", format!("Expected more tokens")))?;

                if peek == &Token::AngleBracketClose {
                    value.pop_front();
                    return Ok(TypeCreation::Touple(exps));
                }

                exps.push(error!(Exp::try_from(&mut *value), "TypeCreation")?)
            }
        }

        let name = error!(NamespacedType::try_from(&mut *value), "TypeCreation")?;

        let ret = match value.first() {
            Some(Token::CurlyOpen) => {
                value.pop_front();
                let mut fields = vec![];

                loop {
                    let peek = value
                        .first()
                        .ok_or(error!("TypeCreation", format!("Expected more tokens")))?;

                    if peek == &Token::CurlyClose {
                        value.pop_front();
                        break;
                    }

                    let next = value.pop_front_err("TypeCreation")?;
                    let name = if let Token::Identifier(iden) = next {
                        iden
                    } else {
                        return Err(error!(
                            "TypeCreation",
                            format!("Expected identifier, got {next:#?}")
                        ));
                    };

                    let next = value.pop_front_err("TypeCreation")?;
                    if next != Token::Keyword(Keywords::Arrow) {
                        return Err(error!(
                            "TypeCreation",
                            format!("Expected arrow keyword, got {next:#?}")
                        ));
                    }

                    let exp = error!(Exp::try_from(&mut *value), "TypeCreation")?;

                    fields.push((name, exp));
                }

                Ok(Self::Struct(name, fields))
            }
            Some(Token::ParenClose) => Ok(Self::Simple(name)),
            Some(_) if paren => {
                let mut args = vec![];

                loop {
                    let peek = value
                        .first()
                        .ok_or(error!("TypeCreation", format!("Expected more tokens")))?;

                    if peek == &Token::ParenClose {
                        value.pop_front();
                        break;
                    }

                    args.push(error!(Exp::try_from(&mut *value), "TypeCreation")?)
                }

                return Ok(Self::Vars(name, args));
            }
            _ => Ok(Self::Simple(name)),
        };

        if paren {
            let next = value.pop_front_err("TypeCreation")?;
            if next != Token::ParenClose {
                return Err(error!(
                    "TypeCreation",
                    format!("Expected parenClose, got {next:#?}")
                ));
            }
        }

        ret
    }
}

impl ToString for TypeCreation {
    fn to_string(&self) -> String {
        match self {
            TypeCreation::Simple(name) => format!("{}", name.to_string()),
            TypeCreation::Vars(name, args) => format!(
                "{}({})",
                name.to_string(),
                &args.iter().fold(String::new(), |str, arg| {
                    format!("{str}, {}", arg.to_string())
                })[2..]
            ),
            TypeCreation::Struct(name, fields) => {
                format!(
                    "{}{{{}}}",
                    name.to_string(),
                    &fields.iter().fold(String::new(), |str, field| {
                        format!("{str}, {}:{}", field.0, field.1.to_string())
                    })[2..]
                )
            }
            TypeCreation::Touple(exps) => format!(
                "({})",
                &if exps.is_empty() {
                    format!(", ")
                } else {
                    exps.iter().fold(String::new(), |str, exp| {
                        format!("{str}, {}", exp.to_string())
                    })
                }[2..]
            ),
        }
    }
}

#[derive(Debug)]
pub enum Infix {
    Add(Exp, Exp),
    Subtract(Exp, Exp),
    Multiply(Exp, Exp),
    Divide(Exp, Exp),
    Equality(Exp, Exp),
    NotEquality(Exp, Exp),
    And(Exp, Exp),
    Or(Exp, Exp),
    Less(Exp, Exp),
    LessEq(Exp, Exp),
    Greater(Exp, Exp),
    GreaterEq(Exp, Exp),
    Xor(Exp, Exp),
    BitwiseAnd(Exp, Exp),
    BitwiseOr(Exp, Exp),
    LeftShift(Exp, Exp),
    RightShift(Exp, Exp),
}

impl TryFrom<&mut Parser> for Infix {
    type Error = ParserError;

    fn try_from(value: &mut Parser) -> Result<Self, Self::Error> {
        Ok(
            match value
                .first()
                .ok_or(error!("Infix", format!("Expected more tokens")))?
            {
                Token::Char('+') => {
                    value.pop_front();
                    Self::Add(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::Char('-') => {
                    value.pop_front();
                    Self::Subtract(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::Char('*') => {
                    value.pop_front();
                    Self::Multiply(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::Slash => {
                    value.pop_front();
                    if value.first() == Some(&Token::Char('=')) {
                        value.pop_front();
                        Self::NotEquality(
                            error!(Exp::try_from(&mut *value), "Infix")?,
                            error!(Exp::try_from(&mut *value), "Infix")?,
                        )
                    } else {
                        Self::Divide(
                            error!(Exp::try_from(&mut *value), "Infix")?,
                            error!(Exp::try_from(&mut *value), "Infix")?,
                        )
                    }
                }
                Token::Char('=') => {
                    value.pop_front();
                    Self::Equality(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::Keyword(Keywords::And) => {
                    value.pop_front();
                    Self::And(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::Keyword(Keywords::Or) => {
                    value.pop_front();
                    Self::Or(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::Keyword(Keywords::Xor) => {
                    value.pop_front();
                    Self::Xor(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::Keyword(Keywords::BitwiseAnd) => {
                    value.pop_front();
                    Self::BitwiseAnd(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::Keyword(Keywords::BitwiseOr) => {
                    value.pop_front();
                    Self::BitwiseOr(
                        error!(Exp::try_from(&mut *value), "Infix")?,
                        error!(Exp::try_from(&mut *value), "Infix")?,
                    )
                }
                Token::AngleBracketOpen => {
                    value.pop_front();
                    match value.first() {
                        Some(&Token::Char('=')) => {
                            value.pop_front();
                            Self::LessEq(
                                error!(Exp::try_from(&mut *value), "Infix")?,
                                error!(Exp::try_from(&mut *value), "Infix")?,
                            )
                        }
                        Some(&Token::AngleBracketOpen) => {
                            value.pop_front();
                            Self::LeftShift(
                                error!(Exp::try_from(&mut *value), "Infix")?,
                                error!(Exp::try_from(&mut *value), "Infix")?,
                            )
                        }
                        _ => Self::Less(
                            error!(Exp::try_from(&mut *value), "Infix")?,
                            error!(Exp::try_from(&mut *value), "Infix")?,
                        ),
                    }
                }
                Token::AngleBracketClose => {
                    value.pop_front();
                    match value.first() {
                        Some(&Token::Char('=')) => {
                            value.pop_front();
                            Self::GreaterEq(
                                error!(Exp::try_from(&mut *value), "Infix")?,
                                error!(Exp::try_from(&mut *value), "Infix")?,
                            )
                        }
                        Some(&Token::AngleBracketClose) => {
                            value.pop_front();
                            Self::RightShift(
                                error!(Exp::try_from(&mut *value), "Infix")?,
                                error!(Exp::try_from(&mut *value), "Infix")?,
                            )
                        }
                        _ => Self::Greater(
                            error!(Exp::try_from(&mut *value), "Infix")?,
                            error!(Exp::try_from(&mut *value), "Infix")?,
                        ),
                    }
                }
                token => {
                    return Err(error!(
                        "Infix",
                        format!("Token isn't infix, got {token:#?}")
                    ))
                }
            },
        )
    }
}

impl ToString for Infix {
    fn to_string(&self) -> String {
        match self {
            Infix::Add(left, right) => format!("({}) + ({})", left.to_string(), right.to_string()),
            Infix::Subtract(left, right) => {
                format!("({}) - ({})", left.to_string(), right.to_string())
            }
            Infix::Multiply(left, right) => {
                format!("({}) * ({})", left.to_string(), right.to_string())
            }
            Infix::Divide(left, right) => {
                format!("({}) / ({})", left.to_string(), right.to_string())
            }
            Infix::Equality(left, right) => {
                format!("({}) == ({})", left.to_string(), right.to_string())
            }
            Infix::NotEquality(left, right) => {
                format!("({}) != ({})", left.to_string(), right.to_string())
            }
            Infix::And(left, right) => format!("({}) && ({})", left.to_string(), right.to_string()),
            Infix::Or(left, right) => format!("({}) || ({})", left.to_string(), right.to_string()),
            Infix::Less(left, right) => format!("({}) < ({})", left.to_string(), right.to_string()),
            Infix::LessEq(left, right) => {
                format!("({}) <= ({})", left.to_string(), right.to_string())
            }
            Infix::Greater(left, right) => {
                format!("({}) > ({})", left.to_string(), right.to_string())
            }
            Infix::GreaterEq(left, right) => {
                format!("({}) >= ({})", left.to_string(), right.to_string())
            }
            Infix::Xor(left, right) => format!("({}) ^ ({})", left.to_string(), right.to_string()),
            Infix::BitwiseAnd(left, right) => {
                format!("({}) & ({})", left.to_string(), right.to_string())
            }
            Infix::BitwiseOr(left, right) => {
                format!("({}) | ({})", left.to_string(), right.to_string())
            }
            Infix::LeftShift(left, right) => {
                format!("({}) << ({})", left.to_string(), right.to_string())
            }
            Infix::RightShift(left, right) => {
                format!("({}) >> ({})", left.to_string(), right.to_string())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::snapshot;

    snapshot!(test_calling, Exp::try_from, "calling.lt");
    snapshot!(test_calling_rust, Exp::try_from, "calling.lt", rust);
}
