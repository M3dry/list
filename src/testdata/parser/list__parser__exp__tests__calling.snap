---
source: src/parser/exp.rs
expression: "contents.lines().filter_map(|line|\n                if line != \"\" {\n                        Some(format!(\"{line}\\n{}\",\n                                |parser| format!(\"{:#?}\", Exp::try_from (parser))\n                                (& mut Parser :: new(line.parse().unwrap()))))\n                    } else { None }).collect::<Vec<String>>().join(\"\\n\\n\")"
---
(+ 4 1)
Ok(
    Infix(
        Add(
            Literal(
                Int(
                    false,
                    4,
                ),
            ),
            Literal(
                Int(
                    false,
                    1,
                ),
            ),
        ),
    ),
)

(= 10 xy)
Ok(
    Infix(
        Equality(
            Literal(
                Int(
                    false,
                    10,
                ),
            ),
            Variable(
                "xy",
            ),
        ),
    ),
)

(func 100 (+ 4 10))
Ok(
    FuncCall(
        Variable(
            "func",
        ),
        [
            Literal(
                Int(
                    false,
                    100,
                ),
            ),
            Infix(
                Add(
                    Literal(
                        Int(
                            false,
                            4,
                        ),
                    ),
                    Literal(
                        Int(
                            false,
                            10,
                        ),
                    ),
                ),
            ),
        ],
    ),
)

(= (func 69 (+ 1 10)) (func 10 100))
Ok(
    Infix(
        Equality(
            FuncCall(
                Variable(
                    "func",
                ),
                [
                    Literal(
                        Int(
                            false,
                            69,
                        ),
                    ),
                    Infix(
                        Add(
                            Literal(
                                Int(
                                    false,
                                    1,
                                ),
                            ),
                            Literal(
                                Int(
                                    false,
                                    10,
                                ),
                            ),
                        ),
                    ),
                ],
            ),
            FuncCall(
                Variable(
                    "func",
                ),
                [
                    Literal(
                        Int(
                            false,
                            10,
                        ),
                    ),
                    Literal(
                        Int(
                            false,
                            100,
                        ),
                    ),
                ],
            ),
        ),
    ),
)

((lambda (x y z) (and (= x y) (= x z))) 10 20 10)
Ok(
    FuncCall(
        Lambda(
            Lambda {
                args: Args(
                    [
                        "x",
                        "y",
                        "z",
                    ],
                ),
                body: Infix(
                    And(
                        Infix(
                            Equality(
                                Variable(
                                    "x",
                                ),
                                Variable(
                                    "y",
                                ),
                            ),
                        ),
                        Infix(
                            Equality(
                                Variable(
                                    "x",
                                ),
                                Variable(
                                    "z",
                                ),
                            ),
                        ),
                    ),
                ),
            },
        ),
        [
            Literal(
                Int(
                    false,
                    10,
                ),
            ),
            Literal(
                Int(
                    false,
                    20,
                ),
            ),
            Literal(
                Int(
                    false,
                    10,
                ),
            ),
        ],
    ),
)
